sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,488 @@
+complexity.
+
+Before compilation of device image, to execute a kernel RT checks if the image
+contains any import information in its properties and if it does, then RT
+performs device images collection in order to resolve dependencies.
+
+Ids of all needed symbol sets are found by iterating through
+`m_SymbolSets` map, i.e. iterating through all available OS modules without
+predefined order and searching for first unresolved symbol in list of imports
+of target device image. Once device image that contains first symbol is
+met, remaining exported symbols are checked in found image. If
+they match some imported symbols then these matched symbols will be marked as
+resolved. The procedure repeats until all imported symbols are marked as
+resolved.
+
+#### Program caching
+
+Existing support for device code caching can be re-used to cache
+dynamically linked programs with slight changes.
+
+##### In-memory cache
+
+The existing mechanism of caching can be re-used in presence of dynamic
+linking. Example:
+```
+// Application
+SYCL_EXTERNAL void LibFunc1();
+SYCL_EXTERNAL void LibFunc2();
+
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for<InternalKernel>( ... )
+}); // 1. Device Image is compiled and linked into a program and saved in cache
+    // 2. Prepared program is used to enqueue kernel
+
+Q.submit([&](cl::sycl::handler &CGH) {
+handler.parallel_for([] { LibFunc1(); }); // Prepared program is used to enqueue kernel
+});
+
+// Library
+SYCL_EXTERNAL void LibFunc1() {
+// ...
+}
+
+```
+In current cache structure the programs map's key consists of four components:
+kernel set id, specialization constants values, the device this program is built
+for, build options id. In this example Id of kernel set where application's
+kernels can be used to access program cache. However when shared library
+defines kernels and these kernels are run by the application unchanged cache
+structure may lead to double compilation of the same code. Example:
+```
+// Application
+SYCL_EXTERNAL void LibFunc();
+
+Q.submit([&](cl::sycl::handler &CGH) {
+handler.parallel_for([] { LibFunc(); });  // Device image for library is compiled
+                                          // and linked together with device
+                                          // image for application, i.e.
+                                          // LibFunc1 and ExternalKernel exist
+                                          // in prepared state
+});
+// ...
+EnqueueLibraryKernel(Q); // If cache mechanism is not changed, this line will
+                         // lead to second compilation of ExternalKernel and
+                         // LibFunc1
+
+// Library
+SYCL_EXTERNAL void LibFunc1() {
+// ...
+}
+
+EnqueueLibraryKernel(queue) {
+  queue.submit(parallel_for<ExternalKernel>(...));
+}
+```
+Such case can be optimized by bringing nesting into cache keys structure.
+Kernel set id can be found for each kernel using its name and OS module it is
```

Comment:
BTW.
OS module is likely will be removed together with kernel set id (sycl/source/detail/program_manager/program_manager.cpp:1229). I think we will just update the design accordingly once it's done.